### PR TITLE
Support movie file uploads, more content-types and server improvements.

### LINKED ISSUE #9: InstantShare client should support movie file uploads (from disk).
The server was designed with this in mind, so it's possible to stream large files as they're being uploaded.

## Desired behavior

When clicking "Instant Share", if the clipboard contains a file, follow these steps:

1. Send the initial request to the InstantShare server, providing the file extension and getting back a link where the file can be accessed once the upload begins.
2. Stream the file content to upload it.

### LINKED ISSUE #12: server: Issues serving content (while uploading, and otherwise).
I'm able to quite consistently reproduce the error when running InstantShare server locally with first download attempt failing if it happens while the upload hasn't finished.

Using Chrome's network tab is helpful in visualizing the problem:

image

The error is mismatch between the response length and content-length value.

I'm also seeing consistent difference in serving a large (3~ MB) file, even after it finished uploading. This might be chunked transmission, or maybe there's something else at play - I'd have to investigate further and compare with how a plain [`http.FileServer`](http://godoc.org/net/http#FileServer) or [`http.ServeContent`](http://godoc.org/net/http#ServeContent) behaves. **Edit:** I've done so below, and there _is_ a difference.

See the 3~ MB movie file served from InstantShare server:

image

It doesn't play in browser.

And the same file accessed from the filesystem via file:// does play in browser:

image

And the same file served using [`http.FileServer`](http://godoc.org/net/http#FileServer) also plays in browser successfully:

image

My conclusion is that the current serving code is subtly not completely correct, and we'll need to fix it to get videos to play and for images to always load on first try (even while uploading). I can take a stab at this later on. I am also seeing opportunity to clean up some code and make it more [idiomatic Go](https://code.google.com/p/go-wiki/wiki/CodeReviewComments) and shorter. Plus there are changes regarding file extensions/content type I need to make (already made locally) to support movie file uploads.

### PR DESCRIPTION
**Pull request by shurcooL**
Resolves #9.
Fixes #12.

~~Not ready for merge, this PR is to track development.~~ Ready for review and merge now.

Requires `trayhost` on [`clipboard-files`](https://github.com/shurcooL/trayhost/tree/feature/clipboard-files) branch.

### CODE REVIEW

**Comment by shurcooL** [comment]
Okay, this is ready for review and merge.

I've reviewed it. There are quite a few minor todos, but I will say let's merge as is because it is functional and the remaining todos can be addressed later as needed. Since this has few users, it doesn't make sense to spend too much time now.

LGTM. Can be merged after https://github.com/shurcooL/trayhost/pull/5 is merged.

File: client/main.go [hunk 1]
```diff
@@ -25,33 +24,54 @@ var debugFlag = flag.Bool("debug", false, "Adds menu items for debugging purpose
...
-
-	img, err := trayhost.GetClipboardImage()
-	if err != nil {
-		log.Println(err)
-		return
-	}
-
 	// Convert image to desired destination format (currently, always PNG).
+	// TODO: Maybe not do this for files? What if it's a jpeg.
+	var extension = "png"
```
Comment (pavben) [inline client/main.go]: Don't default. Set it in the switch.

Comment (shurcooL) [inline client/main.go]: Done in fcf1ed3.

File: server/activefiles.go [hunk 2]
```diff
@@ -160,15 +147,13 @@ func (self *ActiveFileManager) Upload(fileName string, fileData io.ReadCloser, c
 	defer func() {
 		fileWriter.Close()
 
+		// TODO: What error is being checked here? It's very unclear.
```
Comment (pavben) [inline server/activefiles.go]: As explained.

File: client/main.go [hunk 3]
```diff
@@ -25,33 +24,54 @@ var debugFlag = flag.Bool("debug", false, "Adds menu items for debugging purpose
...
 
-	_, err := trayhost.GetClipboardImage()
+	cc, err := trayhost.GetClipboardContent()
 	if err != nil {
 		return false
 	}
 
-	return true
+	switch {
+	case len(cc.Files) == 1 && filepath.Ext(cc.Files[0]) == ".png": // Single .png file.
```
Comment (pavben) [inline client/main.go]:
```go
if len(cc.Files) == 1 {
  ext = ...; // "png" or "mov" without the dot
  readFile(cc.Files[0], ext)
} else if cc.Image.Kind != "" {
  ...
}
```

File: server/activefiles.go (hunk 2 above)
Comment (shurcooL) [inline server/activefiles.go]: Done in c04ba20.

**Pull request merged**